File: FitMealAI/Core/Services/PaymentOptionsService.swift
//
//  PaymentOptionsService.swift
//  FitMealAI
//
//  Calls /api/payments/options to find out which payment methods are
//  available for the current user. The server side computes this from:
//    - app_settings.aba_payment.enabled        (admin toggle)
//    - app_settings.aba_payment.allowed_regions (country allow-list)
//    - the IP-resolved country of this request  (Vercel header)
//
//  We use this to hide the "Pay with ABA" button when the user is
//  outside Cambodia (or wherever the admin has restricted it to), and
//  to know whether the KHQR provider is configured at all.
//

import Foundation

struct PaymentOptions: Decodable, Equatable {
    struct ABAOption: Decodable, Equatable {
        let enabled: Bool
        let allowed_regions: [String]
        let available_for_user: Bool
    }
    struct KhqrOption: Decodable, Equatable {
        let available: Bool
        let active_providers: [String]
    }

    let aba_payment: ABAOption
    let khqr_payment: KhqrOption
    let detected_country: String?

    /// Pessimistic default used when the network call fails. Hides the
    /// ABA button so we don't show a region-locked feature outside its
    /// region; KHQR stays visible because the worst case there is a
    /// transient gateway error.
    static let unavailable = PaymentOptions(
        aba_payment: .init(enabled: false, allowed_regions: [], available_for_user: false),
        khqr_payment: .init(available: true, active_providers: []),
        detected_country: nil,
    )
}

@MainActor
final class PaymentOptionsService {

    private let config: FitMealConfig

    init(config: FitMealConfig) {
        self.config = config
    }

    func fetch() async -> PaymentOptions {
        guard let baseURL = config.apiBaseURL else {
            return .unavailable
        }
        let url = baseURL.appendingPathComponent("api/payments/options")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.cachePolicy = .reloadIgnoringLocalCacheData

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                return .unavailable
            }
            return try JSONDecoder().decode(PaymentOptions.self, from: data)
        } catch {
            return .unavailable
        }
    }
}

File: android/app/src/main/java/com/fitmealai/data/PaymentOptionsService.kt
package com.fitmealai.data

import com.fitmealai.config.AppConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Calls /api/payments/options to find out which payment methods are
 * available for the current user. The server side computes this from:
 *   - app_settings.aba_payment.enabled        (admin toggle)
 *   - app_settings.aba_payment.allowed_regions (country allow-list)
 *   - the IP-resolved country of this request  (Vercel header)
 *
 * Mirrors the iOS [PaymentOptionsService]. We use this to hide the
 * "Pay with ABA" button on the paywall when the user is outside the
 * Cambodia allow-list.
 */
class PaymentOptionsService(private val config: AppConfig = AppConfig()) {

    /**
     * Snapshot of `/api/payments/options`. Only the booleans are
     * load-bearing; `detectedCountry` is informational.
     */
    data class Options(
        val abaEnabled: Boolean,
        val abaAllowedRegions: List<String>,
        val abaAvailableForUser: Boolean,
        val khqrAvailable: Boolean,
        val activeKhqrProviders: List<String>,
        val detectedCountry: String?,
    ) {
        companion object {
            /**
             * Pessimistic default used when the network call fails: hides
             * the ABA button (we don't want to show a region-locked feature
             * outside its region) but keeps KHQR visible since the worst
             * case there is a transient gateway error.
             */
            val Unavailable = Options(
                abaEnabled = false,
                abaAllowedRegions = emptyList(),
                abaAvailableForUser = false,
                khqrAvailable = true,
                activeKhqrProviders = emptyList(),
                detectedCountry = null,
            )
        }
    }

    suspend fun fetch(): Options {
        if (config.apiBaseUrl.isBlank()) return Options.Unavailable
        val url = "${config.apiBaseUrl.trimEnd('/')}/api/payments/options"
        return withContext(Dispatchers.IO) {
            try {
                val connection = (URL(url).openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    connectTimeout = 10_000
                    readTimeout = 10_000
                    setRequestProperty("Accept", "application/json")
                }
                if (connection.responseCode !in 200..299) {
                    return@withContext Options.Unavailable
                }
                val text = connection.inputStream.bufferedReader().use { it.readText() }
                val json = JSONObject(text)
                val aba = json.optJSONObject("aba_payment") ?: JSONObject()
                val khqr = json.optJSONObject("khqr_payment") ?: JSONObject()
                Options(
                    abaEnabled = aba.optBoolean("enabled", false),
                    abaAllowedRegions = aba.optJSONArray("allowed_regions")?.let { arr ->
                        (0 until arr.length()).mapNotNull { arr.optString(it).ifBlank { null } }
                    } ?: emptyList(),
                    abaAvailableForUser = aba.optBoolean("available_for_user", false),
                    khqrAvailable = khqr.optBoolean("available", true),
                    activeKhqrProviders = khqr.optJSONArray("active_providers")?.let { arr ->
                        (0 until arr.length()).mapNotNull { arr.optString(it).ifBlank { null } }
                    } ?: emptyList(),
                    detectedCountry = json.optString("detected_country").ifBlank { null },
                )
            } catch (_: Throwable) {
                Options.Unavailable
            }
        }
    }
}
